### PR TITLE
feat(streams): server-side audience filtering

### PR DESCRIPTION
**Pull request by mhenrixon**
Stacked on #84 (replay). **Base: `feature/turbo-streams-phase-9-replay`**.

## Summary

Per-connection broadcast filtering. Users register predicate labels at boot, then opt in to filtering on individual broadcasts:

```ruby
# config/initializers/pgbus_streams.rb
Pgbus::Streams.filters.register(:admin_only) { |user| user.admin? }

# Anywhere in the app
Pgbus.stream("ops").broadcast(html, visible_to: :admin_only)
```

Connections get a context object from the StreamApp's authorize hook (return any non-boolean truthy value, typically a User model). The Dispatcher evaluates the named predicate against each connection's context and only delivers to matching connections.

## Why this matters

Before this PR, restricting a broadcast to a subset of subscribers meant either:

1. **Broadcasting to per-user streams** — explodes to N streams per broadcast, doesn't scale, and fragments the LISTEN channel space
2. **Broadcasting to everyone and filtering client-side** — leaks data, defeats the point of authorization

Now: one stream, one broadcast, server-side enforcement, zero data leakage.

## Failure semantics (deliberate)

| Case | Behavior | Why |
|---|---|---|
| `visible_to: nil` (no label) | Deliver to all | Backward compatible — unfiltered broadcasts work as before |
| Label not in registry | **Fail-OPEN** + warning log | A typo shouldn't silently drop every broadcast to zero subscribers; the developer needs to see the error in logs |
| Predicate raises | **Fail-CLOSED** | A buggy predicate that crashes is treated as "deny" — better to drop one broadcast than to leak private data on an exception path |

## Integration test: 2 clients, 2 contexts, 2 broadcasts

`spec/integration/streams/audience_spec.rb`:

1. Admin client connects with authorize context `{ role: 'admin' }`
2. Viewer client connects with authorize context `{ role: 'viewer' }`
3. Public broadcast: `stream.broadcast(html)`
4. Admin-only broadcast: `stream.broadcast(html, visible_to: :admin_only)`
5. Admin client receives both, viewer receives only the public one

Real Puma + real PGMQ + real LISTEN/NOTIFY + real Filter registry. 1.3 seconds.

## Pre-existing latent bug fixed as a side effect

The engine's `pgbus.streams.turbo_broadcastable` initializer was calling `Pgbus::Streams.install_turbo_broadcastable_patch!` without first referencing `Pgbus::Streams::TurboBroadcastable` as a constant, which meant Zeitwerk never loaded the file that defines the method. Any spec that boots Rails before any other code touches the constant (e.g. `spec/pgbus/web/locale_spec.rb`) crashed at boot with `NoMethodError`.

This was already broken on `main` — confirmed by stashing my changes and reproducing. Fix: assign the constant to a `_autoload_trigger` local before the method call so Zeitwerk loads the file. Underscore prefix keeps RuboCop's Lint/Void from deleting the line during autocorrect.

## Tests: crossed the 1000 mark

- **985 unit + 15 integration = 1000 passing, 0 failures**
- 11 new Filters unit examples
- 5 new Dispatcher unit examples covering the filter-application path
- 1 new integration test covering the full pipeline
- RuboCop clean across 267 files

## Test plan

- [x] `bundle exec rspec spec/pgbus/streams/filters_spec.rb` — 11 examples
- [x] `bundle exec rspec spec/pgbus/web/streamer/dispatcher_spec.rb` — 18 examples (13 existing + 5 new)
- [x] `bundle exec rspec spec/pgbus/` — 985 examples, full unit regression
- [x] `PGBUS_DATABASE_URL=... bundle exec rspec spec/integration/streams/audience_spec.rb` — 1 end-to-end
- [x] `PGBUS_DATABASE_URL=... bundle exec rspec spec/integration/streams/` — 15 examples, all streams integration tests
- [x] `bundle exec rubocop` clean across 267 files
- [ ] Reviewer: confirm the failure semantics (fail-open on unknown, fail-closed on raise) match what you want
- [ ] Reviewer: read the engine fix — `_autoload_trigger = Pgbus::Streams::TurboBroadcastable` is unusual but necessary

## What's next in the stack

- #81 ✅ — Transactional broadcasts
- #83 ✅ — Falcon server support
- #84 ✅ — `broadcasts_with_replay`
- **#85 (this PR)** — Server-side audience filtering
- #86 — Presence as a stream
- #87 — Performance sweep + benchmarks
- #88 — Docs refresh

## Refs

- Phase 0 design doc §12 roadmap item 6 — explicitly deferred in v1
- Earlier 'Tier 2' brainstorm — listed as a high-value differentiator after transactional broadcasts

## Summary by CodeRabbit

* **New Features**
  * Server-side audience filtering for real-time streams with named filter predicates.
  * Broadcasts can include a visible_to label so only matching subscribers receive a message.
  * Per-connection authorize may return a context value passed to predicates for fine-grained delivery.

* **Behavior / Reliability**
  * Unknown labels or predicate errors are treated as fail-closed (message dropped) and logged.
  * Filters are process-local (per worker/reactor).

* **Documentation**
  * README updated with configuration, usage, and semantics.

* **Tests**
  * Integration and unit tests added for audience filtering and edge cases.

### CODE REVIEW

File: lib/pgbus/web/streamer/dispatcher.rb [hunk 1]
```diff
@@ -130,8 +145,15 @@ def handle_wake(msg)
...
-          in_flight_pairs.each { |(_conn, buffer)| buffer.concat(envelopes) }
+          # Each connection gets a per-connection filtered subset. We
+          # can't pre-filter once because different connections have
+          # different authorize contexts.
+          registered.each do |conn|
+            safe_enqueue(conn, visible_envelopes_for(envelopes, conn))
+          end
+          in_flight_pairs.each do |(conn, buffer)|
+            buffer.concat(visible_envelopes_for(envelopes, conn))
+          end
```
[the remark this answers is not in the document]
Comment (mhenrixon) [reply lib/pgbus/web/streamer/dispatcher.rb:177]: Fixed in 40bf3fb — real bug. Added a per-connection `@scanned_cursor` hash on Dispatcher that tracks the highest `msg_id` examined (not delivered) per connection. `minimum_cursor` now uses `max(scanned_cursor, last_msg_id_sent)` so a run of audience-filtered messages no longer pins the replay window. `advance_scanned_cursor` is called after each `visible_envelopes_for` pass in `handle_wake`, scrubbed from `handle_disconnect` and the dead-before-register + rescue paths in `handle_connect`. `Connection#last_msg_id_sent` still drives the client-facing Last-Event-ID so SSE clients see the exact same cursor semantics — the scan cursor is Dispatcher-internal only.

New dispatcher spec covers the scenario you described: a batch of 3 admin-only messages followed by a public message; without the scan cursor the viewer would never see the public message because `read_after(after_id: 0)` would keep returning the same filtered batch.

File: lib/pgbus/web/streamer/dispatcher.rb [hunk 2]
```diff
@@ -268,15 +291,32 @@ def unwrap_stream_envelope(envelope)
...
+        # context. Envelopes without a visible_to label pass through
+        # unchanged; envelopes with a label are evaluated via the
+        # Filters registry. Envelopes that predate the StreamEnvelope
+        # refactor (plain ReadAfter::Envelope with no visible_to) also
+        # pass through.
+        def visible_envelopes_for(envelopes, connection)
+          envelopes.select do |envelope|
+            label = envelope.respond_to?(:visible_to) ? envelope.visible_to : nil
+            @filters.visible?(label, connection.context)
+          end
```
[the remark this answers is not in the document]
Comment (mhenrixon) [reply lib/pgbus/web/streamer/dispatcher.rb:361]: Fixed in 40bf3fb. Flipped `Filters#visible?` from fail-open to fail-closed on unknown labels. You're right — the old comment explicitly documented fail-open as "better debugging" but that's the wrong trade-off for a data-isolation feature. A typo like `visible_to: :admin_ony` would silently turn a restricted broadcast into a public one, which is exactly the opposite of what the developer intended. Fail-closed with a warning log is still debuggable ("why are no subscribers receiving my broadcast?" → check the log) and doesn't leak data on a silent regression.

Did NOT add validation in `Stream#broadcast` — the broadcast runs in the application process which may not have all filters registered yet (or may have a different set than the streamer process). Validating at broadcast-time would couple the two processes, and the fail-closed runtime behavior is sufficient on its own.

Updated the README's failure-semantics bullet, the filters_spec, and the dispatcher_spec's previously-named "fail-opens on unknown label" test (now "fail-closes on unknown filter label").

File: spec/integration/streams/audience_spec.rb [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+  before do
+    Pgbus.client.ensure_stream_queue(stream_name)
+    streamer.start
+  end
+
+  after do
+    streamer.shutdown!
+    admin_harness.shutdown if defined?(@admin_started)
+    viewer_harness.shutdown if defined?(@viewer_started)
+  end
```
[the remark this answers is not in the document]
Comment (mhenrixon) [reply spec/integration/streams/audience_spec.rb:90]: Fixed in 40bf3fb. Applied the same captured-harness teardown pattern that PR #83 round 2 normalised across all streams integration specs. `connect_admin` / `connect_viewer` now assign `@booted_admin` / `@booted_viewer` to the harness *after* the lazy `let` successfully returns, and the `after` hook does `@booted_admin&.shutdown` / `@booted_viewer&.shutdown` — so a boot failure leaves the ivars nil and teardown is a safe no-op instead of re-triggering the lazy let.

File: spec/pgbus/streams/filters_spec.rb [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+      # Audience filtering is a data-isolation feature. Failing open
+      # on a typo'd or renamed label would turn a restricted broadcast
+      # into a public one — the opposite of what the developer wanted.
+      # The warning log is loud enough that typos still get noticed
+      # in dev ("why are no subscribers receiving my broadcast?" →
+      # check the log).
+      warnings = []
+      filters_with_logger = described_class.new(logger: double("logger", warn: ->(&b) { warnings << b.call }))
+      expect(filters_with_logger.visible?(:nope, { any: :context })).to be false
+    end
```
[the remark this answers is not in the document]
Comment (mhenrixon) [reply spec/pgbus/streams/filters_spec.rb:72]: Fixed in b808c4e. Both the unknown-label and raising-predicate fail-closed tests now assert the `logger.warn`/`logger.error` block fires and the captured message matches a regex. The log is the developer-facing signal that makes fail-closed debuggable; without a test covering it, a future refactor could silently drop the log and regress the "typos are loud" guarantee.

File: spec/pgbus/streams/filters_spec.rb [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+    it "returns true when the label is nil (no filter applied)" do
+      expect(filters.visible?(nil, { any: :context })).to be true
+    end
+
+    it "gracefully handles a raising filter by logging and returning false (fail-closed on exception)" do
+      # If the predicate itself raises, we fail CLOSED — better to drop
+      # one broadcast than to show private data to the wrong user.
+      filters.register(:broken) { |_ctx| raise "boom" }
+      expect(filters.visible?(:broken, {})).to be false
+    end
```
[the remark this answers is not in the document]
Comment (mhenrixon) [reply spec/pgbus/streams/filters_spec.rb:91]: Fixed in b808c4e — same fix as the unknown-label test, applied to the raising-predicate test too. Both fail-closed tests now capture the log block and assert the message.
